File: Backend (Springboot)/Annapurna/src/main/java/com/annapurna/service/UserService.java
package com.annapurna.service;

import com.annapurna.domain.*;
import com.annapurna.dto.UserDTO;
import com.annapurna.dto.AddressDTO;
import com.annapurna.repository.*;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final TiffinRepository tiffinRepository;
    private final OrderRepository orderRepository;
    private final AddressRepository addressRepository;

    
    @Transactional
    public User registerUser(UserDTO dto) {
        
        if (userRepository.findByEmail(dto.getEmail()).isPresent()) {
            throw new RuntimeException("Email already exists");
        }

        User user = User.builder()
                .fullName(dto.getFullName())
                .email(dto.getEmail())
                .password(passwordEncoder.encode(dto.getPassword()))
                .role(dto.getRole())
                .mobileNumber(dto.getMobileNumber())
                .build();

        User savedUser = userRepository.save(user);

        
        if (dto.getAddress() != null) {
            AddressDTO addressDTO = dto.getAddress();
            Address address = Address.builder()
                    .street(addressDTO.getStreet())
                    .city(addressDTO.getCity())
                    .state(addressDTO.getState())
                    .pinCode(addressDTO.getPinCode())
                    .country(addressDTO.getCountry())
                    .user(savedUser)
                    .build();
            addressRepository.save(address);
        }

        return savedUser;
    }

    public User getByEmail(String email) {
        return userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found"));
    }

    public User getUserByEmail(String email) {
        return userRepository.findByEmail(email)
                .orElseThrow(() -> new UsernameNotFoundException("User not found"));
    }

    
    @Transactional
    public void deleteVendorById(Long vendorId) {
        User vendor = userRepository.findById(vendorId)
                .orElseThrow(() -> new RuntimeException("Vendor not found with ID: " + vendorId));

        if (vendor.getRole() != Role.VENDOR) {
            throw new RuntimeException("User is not a vendor.");
        }

        
        List<Tiffin> tiffins = tiffinRepository.findByVendor(vendor);
        for (Tiffin t : tiffins) {
            orderRepository.deleteByTiffin(t);
        }

       
        tiffinRepository.deleteAll(tiffins);

        
        addressRepository.deleteByUser(vendor);

        
        userRepository.delete(vendor);
    }
}

File: Backend (DNET)/Annapurna/Services/UserService.cs
using Microsoft.EntityFrameworkCore;
using AnnapurnaAPI.Data;
using AnnapurnaAPI.Models;
using AnnapurnaAPI.DTOs;
using BCrypt.Net;

namespace AnnapurnaAPI.Services
{
    public class UserService
    {
        private readonly ApplicationDbContext _context;

        public UserService(ApplicationDbContext context)
        {
            _context = context;
        }

        public async Task<User> RegisterUser(UserDTO dto)
        {
            var existingUser = await _context.Users.FirstOrDefaultAsync(u => u.Email == dto.Email);
            if (existingUser != null)
            {
                throw new InvalidOperationException("User with this email already exists");
            }

            var hashedPassword = BCrypt.Net.BCrypt.HashPassword(dto.Password);

            if (!Enum.IsDefined(typeof(Role), dto.Role))
            {
                throw new InvalidOperationException("Invalid role provided");
            }

            Role role = (Role)dto.Role;

            var user = new User
            {
                FullName = dto.FullName,
                Email = dto.Email,
                Password = hashedPassword,
                Role = role,
                MobileNumber = dto.MobileNumber,
                Address = dto.Address != null ? new Address
                {
                    Street = dto.Address.Street,
                    City = dto.Address.City,
                    State = dto.Address.State,
                    PinCode = dto.Address.PinCode,
                    Country = dto.Address.Country
                } : null
            };

            _context.Users.Add(user);
            await _context.SaveChangesAsync();

            return user;
        }






        public async Task<User?> GetUserByEmail(string email)
        {
            return await _context.Users
                .Include(u => u.Address)
                .FirstOrDefaultAsync(u => u.Email == email);
        }

        public async Task<User?> GetByEmail(string email)
        {
            return await _context.Users
                .Include(u => u.Address)
                .FirstOrDefaultAsync(u => u.Email == email);
        }

        public async Task<bool> ValidatePassword(string email, string password)
        {
            var user = await GetUserByEmail(email);
            if (user == null) return false;

            return BCrypt.Net.BCrypt.Verify(password, user.Password);
        }

        public async Task DeleteVendorById(long vendorId)
        {
            var vendor = await _context.Users
                .Include(u => u.Tiffins)
                .ThenInclude(t => t.Orders)
                .FirstOrDefaultAsync(u => u.Id == vendorId && u.Role == Role.VENDOR);

            if (vendor != null)
            {
                _context.Users.Remove(vendor);
                await _context.SaveChangesAsync();
            }
        }
    }
}
